File: tests/mdts/lib/rule.py
# Copyright 2014 Midokura SARL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from mdts.lib.resource_base import ResourceBase


FIELDS = [
    'inv_port_group',
    'tp_src',
    'dl_src',
    'dl_src_mask',
    'inv_nw_dst',
    'dl_dst',
    'dl_dst_mask',
    'match_forward_flow',
    'inv_tp_src',
    'match_return_flow',
    'inv_nw_src',
    'out_ports',
    'nw_dst_length',
    'inv_out_ports',
    'position',
    'dl_type',
    'inv_nw_tos',
    'port_group_src',
    'inv_dl_dst',
    'inv_in_ports',
    'jump_chain_name',
    'jump_chain_id',
    'inv_dl_type',
    'inv_tp_dst',
    'chain_id',
    'nw_tos',
    'nw_proto',
    'nw_src_length',
    'in_ports',
    'nw_dst_address',
    'nw_src_address',
    'inv_nw_proto',
    'properties',
    'cond_invert',
    'type',
    'inv_dl_src',
    'tp_dst',
    'flow_action',
    'nat_targets',
    'fragment_policy',
    'target_port'
]


class Rule(ResourceBase):
    """ A class representing a single rule for functional tests. """

    def __init__(self, api, context, data, chain):
        """ Initializes a rule.

        Args:
            api: MidoNet API client object
            context: context for this topology
            data: topology data that represents this resource and below
                  in the hierarchy
            chain_id: Id of the chain this rule belongs to.
        """
        super(Rule, self).__init__(api, context, data)
        self._chain = chain

    def build(self):
        """ Builds MidoNet resource for chain from this data.

            Use the chain ID assigned upon chain creation.
        """
        self._mn_resource = self._chain._mn_resource.add_rule()
        self._mn_resource.chain_id(self._chain._mn_resource.get_id())
        for field in FIELDS:
            if field is 'chain_id' or field not in self._data:
                continue
            if field in ['in_ports', 'out_ports',
                         'port_group_src', 'target_port']:
                self._context.look_up_resource(
                        self._mn_resource, field, self._data[field])
            elif field is 'jump_chain_name':
                self._mn_resource.jump_chain_name(self._data[field])
                if 'jump_chain_id' not in self._data:
                    self._context.look_up_resource(
                            self._mn_resource, 'jump_chain_id',
                            {'chain_name': self._data[field]})
            else:
                getattr(self._mn_resource, field)(self._data[field])
        self._mn_resource.create()

    def destroy(self):
        """ Destroys the rule resource. """
        self._mn_resource.delete()

    def get_id(self):
        """ Returns the rule ID. """
        return self._data.get('id')

    def get_chain_id(self):
        """ Returns the rule chain ID. """
        return self._chain.get_id()

    def get_type(self):
        """ Returns the rule type. """
        return self._data.get('type')

File: cluster/midonet-cluster/src/main/java/org/midonet/client/resource/Rule.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.client.resource;

import java.net.URI;
import java.util.Map;
import java.util.UUID;

import org.midonet.client.WebResource;
import org.midonet.client.dto.DtoRule;
import org.midonet.cluster.services.rest_api.MidonetMediaTypes;

public class Rule extends ResourceBase<Rule, DtoRule> {


    public Rule(WebResource resource, URI uriForCreation, DtoRule r) {
        super(resource, uriForCreation, r,
                MidonetMediaTypes.APPLICATION_RULE_JSON_V2());
    }

    @Override
    public URI getUri() {
        return principalDto.getUri();
    }

    public boolean isCondInvert() {
        return principalDto.isCondInvert();
    }

    public boolean isInvDlDst() {
        return principalDto.isInvDlDst();
    }

    public boolean isInvDlSrc() {
        return principalDto.isInvDlSrc();
    }

    public boolean isInvDlType() {
        return principalDto.isInvDlType();
    }

    public boolean isInvInPorts() {
        return principalDto.isInvInPorts();
    }

    public boolean isInvNwDst() {
        return principalDto.isInvNwDst();
    }

    public boolean isInvNwProto() {
        return principalDto.isInvNwProto();
    }

    public boolean isInvNwSrc() {
        return principalDto.isInvNwSrc();
    }

    public boolean isInvNwTos() {
        return principalDto.isInvNwTos();
    }

    public String getFragmentPolicy() {
        return principalDto.getFragmentPolicy();
    }

    public boolean isInvOutPorts() {
        return principalDto.isInvOutPorts();
    }

    public boolean isInvPortGroup() {
        return principalDto.isInvPortGroup();
    }

    public boolean isInvTpDst() {
        return principalDto.isInvTpDst();
    }

    public boolean isInvTpSrc() {
        return principalDto.isInvTpSrc();
    }

    public boolean isMatchForwardFlow() {
        return principalDto.isMatchForwardFlow();
    }

    public boolean isMatchReturnFlow() {
        return principalDto.isMatchReturnFlow();
    }

    public UUID getChainId() {
        return principalDto.getChainId();
    }

    public String getDlDst() {
        return principalDto.getDlDst();
    }

    public String getDlSrc() {
        return principalDto.getDlSrc();
    }

    public Integer getDlType() {
        return principalDto.getDlType();
    }

    public String getFlowAction() {
        return principalDto.getFlowAction();
    }

    public UUID getId() {
        return principalDto.getId();
    }

    public UUID[] getInPorts() {
        return principalDto.getInPorts();
    }

    public String getJumpChainName() {
        return principalDto.getJumpChainName();
    }

    public UUID getJumpChainId() {
        return principalDto.getJumpChainId();
    }

    public DtoRule.DtoNatTarget[] getNatTargets() {
        return principalDto.getNatTargets();
    }

    public String getNwDstAddress() {
        return principalDto.getNwDstAddress();
    }

    public int getNwDstLength() {
        return principalDto.getNwDstLength();
    }

    public int getNwProto() {
        return principalDto.getNwProto();
    }

    public String getNwSrcAddress() {
        return principalDto.getNwSrcAddress();
    }

    public int getNwSrcLength() {
        return principalDto.getNwSrcLength();
    }

    public int getNwTos() {
        return principalDto.getNwTos();
    }

    public UUID[] getOutPorts() {
        return principalDto.getOutPorts();
    }

    public UUID getPortGroup() {
        return principalDto.getPortGroup();
    }

    public int getPosition() {
        return principalDto.getPosition();
    }

    public String getMeterName() {
        return principalDto.getMeterName();
    }

    public Map<String, String> getProperties() {
        return principalDto.getProperties();
    }

    public DtoRule.DtoRange<Integer> getTpDst() {
        return principalDto.getTpDst();
    }

    public DtoRule.DtoRange<Integer> getTpSrc() {
        return principalDto.getTpSrc();
    }

    public String getType() {
        return principalDto.getType();
    }

    public Rule invNwDst(boolean invNwDst) {
        principalDto.setInvNwDst(invNwDst);
        return this;
    }

    public Rule invNwTos(boolean invNwTos) {
        principalDto.setInvNwTos(invNwTos);
        return this;
    }

    public Rule setFragmentPolicy(String fragmentPolicy) {
        principalDto.setFragmentPolicy(fragmentPolicy);
        return this;
    }

    public Rule chainId(UUID chainId) {
        principalDto.setChainId(chainId);
        return this;
    }

    public Rule position(int position) {
        principalDto.setPosition(position);
        return this;
    }

    public Rule meterName(String meterName) {
        principalDto.setMeterName(meterName);
        return this;
    }

    public Rule invTpDst(boolean invTpDst) {
        principalDto.setInvTpDst(invTpDst);
        return this;
    }

    public Rule invNwSrc(boolean invNwSrc) {
        principalDto.setInvNwSrc(invNwSrc);
        return this;
    }

    public Rule invOutPorts(boolean invOutPorts) {
        principalDto.setInvOutPorts(invOutPorts);
        return this;
    }

    public Rule outPorts(UUID[] outPorts) {
        principalDto.setOutPorts(outPorts);
        return this;
    }

    public Rule nwProto(int nwProto) {
        principalDto.setNwProto(nwProto);
        return this;
    }

    public Rule invDlType(boolean invDlType) {
        principalDto.setInvDlType(invDlType);
        return this;
    }

    public Rule tpDst(DtoRule.DtoRange<Integer> tpDst) {
        principalDto.setTpDst(tpDst);
        return this;
    }

    public Rule dlDst(String dlDst) {
        principalDto.setDlDst(dlDst);
        return this;
    }

    public Rule portGroup(UUID portGroup) {
        principalDto.setPortGroup(portGroup);
        return this;
    }

    public Rule matchReturnFlow(boolean matchReturnFlow) {
        principalDto.setMatchReturnFlow(matchReturnFlow);
        return this;
    }

    public Rule nwDstAddress(String nwDstAddress) {
        principalDto.setNwDstAddress(nwDstAddress);
        return this;
    }

    public Rule invDlDst(boolean invDlDst) {
        principalDto.setInvDlDst(invDlDst);
        return this;
    }

    public Rule invInPorts(boolean invInPorts) {
        principalDto.setInvInPorts(invInPorts);
        return this;
    }

    public Rule dlType(Integer dlType) {
        principalDto.setDlType(dlType);
        return this;
    }

    public Rule natTargets(DtoRule.DtoNatTarget[] natTargets) {
        principalDto.setNatTargets(natTargets);
        return this;
    }

    public Rule invTpSrc(boolean invTpSrc) {
        principalDto.setInvTpSrc(invTpSrc);
        return this;
    }

    public Rule properties(Map<String, String> properties) {
        principalDto.setProperties(properties);
        return this;
    }

    public Rule jumpChainName(String jumpChainName) {
        principalDto.setJumpChainName(jumpChainName);
        return this;
    }

    public Rule jumpChainId(UUID jumpChainId) {
        principalDto.setJumpChainId(jumpChainId);
        return this;
    }

    public Rule matchForwardFlow(boolean matchForwardFlow) {
        principalDto.setMatchForwardFlow(matchForwardFlow);
        return this;
    }

    public Rule nwSrcLength(int nwSrcLength) {
        principalDto.setNwSrcLength(nwSrcLength);
        return this;
    }

    public Rule flowAction(String flowAction) {
        principalDto.setFlowAction(flowAction);
        return this;
    }

    public Rule condInvert(boolean condInvert) {
        principalDto.setCondInvert(condInvert);
        return this;
    }

    public Rule dlSrc(String dlSrc) {
        principalDto.setDlSrc(dlSrc);
        return this;
    }

    public Rule invDlSrc(boolean invDlSrc) {
        principalDto.setInvDlSrc(invDlSrc);
        return this;
    }

    public Rule nwDstLength(int nwDstLength) {
        principalDto.setNwDstLength(nwDstLength);
        return this;
    }

    public Rule invPortGroup(boolean invPortGroup) {
        principalDto.setInvPortGroup(invPortGroup);
        return this;
    }

    public Rule nwTos(int nwTos) {
        principalDto.setNwTos(nwTos);
        return this;
    }

    public Rule nwSrcAddress(String nwSrcAddress) {
        principalDto.setNwSrcAddress(nwSrcAddress);
        return this;
    }

    public Rule inPorts(UUID[] inPorts) {
        principalDto.setInPorts(inPorts);
        return this;
    }

    public Rule tpSrc(DtoRule.DtoRange<Integer> tpSrc) {
        principalDto.setTpSrc(tpSrc);
        return this;
    }

    public Rule invNwProto(boolean invNwProto) {
        principalDto.setInvNwProto(invNwProto);
        return this;
    }

    public Rule type(String type) {
        principalDto.setType(type);
        return this;
    }
}
